src/libexpr/eval.cc

```diff
@@ -716,41 +716,51 @@ Value * ExprPath::maybeThunk(EvalState & state, Env & env)
 
 void EvalState::evalFile(const Path & path_, Value & v)
 {
-    auto path = checkSourcePath(path_);
+    if (path_ == "-") {
+        Expr * e = parseStdin();
+        try {
+            eval(e, v);
+        } catch (Error & e) {
+            addErrorPrefix(e, "while evaluating %1%:\n", "<stdin>");
+            throw;
+        }
+    } else {
+        auto path = checkSourcePath(path_);
 
-    FileEvalCache::iterator i;
-    if ((i = fileEvalCache.find(path)) != fileEvalCache.end()) {
-        v = i->second;
-        return;
-    }
+        FileEvalCache::iterator i;
+        if ((i = fileEvalCache.find(path)) != fileEvalCache.end()) {
+            v = i->second;
+            return;
+        }
 
-    Path path2 = resolveExprPath(path);
-    if ((i = fileEvalCache.find(path2)) != fileEvalCache.end()) {
-        v = i->second;
-        return;
-    }
+        Path path2 = resolveExprPath(path);
+        if ((i = fileEvalCache.find(path2)) != fileEvalCache.end()) {
+            v = i->second;
+            return;
+        }
 
-    printTalkative("evaluating file '%1%'", path2);
-    Expr * e = nullptr;
+        printTalkative("evaluating file '%1%'", path2);
+        Expr * e = nullptr;
 
-    auto j = fileParseCache.find(path2);
-    if (j != fileParseCache.end())
-        e = j->second;
+        auto j = fileParseCache.find(path2);
+        if (j != fileParseCache.end())
+            e = j->second;
 
-    if (!e)
-        e = parseExprFromFile(checkSourcePath(path2));
+        if (!e)
+            e = parseExprFromFile(checkSourcePath(path2));
 
-    fileParseCache[path2] = e;
+        fileParseCache[path2] = e;
 
-    try {
-        eval(e, v);
-    } catch (Error & e) {
-        addErrorPrefix(e, "while evaluating the file '%1%':\n", path2);
-        throw;
-    }
+        try {
```

Comment:
oh, was the bit inserted at the top cloned from this? Does the code lend itself towards handling parsing/caching as appropriate, followed by single block for eval/error handling? Meh, I see some of my nits from earlier were from pre-existing code, sorry.  Ignore if they seem tedious, cleanup of existing code can and probably should be handled orthogonally...